src/main/java/uk/ac/sanger/sccp/stan/service/PermServiceImp.java

```diff
@@ -200,27 +255,74 @@ public void validatePermValues(Collection<String> problems, Collection<PermData>
         if (!addressesNeither.isEmpty()) {
             problems.add("Neither control type nor time specified for the given address: "+addressesNeither);
         }
+        if (!addressesWithBarcodeNotPosCon.isEmpty()) {
+            problems.add("Control barcode specified for address that is not a positive control: "+addressesWithBarcodeNotPosCon);
+        }
     }
 
     /**
      * Records the requested operation and measurements, after validation.
      * @param user the user responsible for the operation
      * @param lw the labware involved in the operation
      * @param permData the perm data to record as measurements
+     * @param controlLabware the labware specified for positive controls, if any
      * @param work the work to associate the operation with, or null
      * @return the operation recorded and the labware
      */
-    public OperationResult record(User user, Labware lw, Collection<PermData> permData, Work work) {
+    public OperationResult record(User user, Labware lw, Collection<PermData> permData, Labware controlLabware, Work work) {
         OperationType opType = opTypeRepo.getByName("Visium permabilisation");
+
+        Operation addControlOp;
+        if (controlLabware!=null) {
+            Address address = permData.stream()
+                    .filter(pd -> pd.getControlBarcode()!=null)
+                    .map(PermData::getAddress)
+                    .findAny()
+                    .orElseThrow();
```

Comment:
Do we need to ensure here that only one address is returned?

Reply:
At this point in the code, findControlLabware() has already checked that only one address has a control barcode.